### DISCUSSION TITLE
How to get full geometry information from Brep when IFCPRODUCTDEFINITIONSHAPE has multiple  IFCSHAPEREPRESENTATION?

### DISCUSSION THREAD

**Discussion by sennhvi** [IfcOpenShell/IfcOpenShell#4146] [General]
**How to get full geometry information from Brep when IFCPRODUCTDEFINITIONSHAPE has multiple  IFCSHAPEREPRESENTATION?**

I have viewed some pages, like trying use settings to show Brep data -- 
 https://blenderbim.org/docs-python/ifcopenshell/geometry_settings.html#disable-triangulation, but I don't know how to use the output and seems the output doesnot contains all vertices I need(what i really care).
 
#2744 gave me some inspiration that maybe I need to use sth like setting up a MODEL_VIEW to get all the geometry data.

When I use code below, I only get geometry data from the first SHAPEREPRESENTATION.
```
import ifcopenshell
import ifcopenshell.geom
import  ifcopenshell.util.shape

test_file = r'a-window.ifc'
ifc_file = ifcopenshell.open(test_file)
window = ifc_file.by_type('IfcBuildingElement')[0]

settings = ifcopenshell.geom.settings()
shape = ifcopenshell.geom.create_shape(settings, window)
matrix = ifcopenshell.util.shape.get_shape_matrix(shape)
location = matrix[:,3][0:3]


verts = shape.geometry.verts
edges = shape.geometry.edges
faces = shape.geometry.faces

```

[a-window.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/13770053/a-window.ifc.txt)

verts is a tuple contains 24 float number which stands for 8 3d points.

Stuck in how to get all verts\edges\faces in Brep. Any idea is welcome:)

**Comment by sennhvi** [comment]
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/31751605/45436e17-8524-495c-be26-0617edcc073d)

The comparison pic shows that the missing geometry is necessary.

**Reply by sennhvi** [reply]
https://blenderbim.org/docs-python/ifcopenshell-python/geometry_processing.html#geometry-iterator use iterator may solve this problem regardless of its MODEL_VIEW or SETTINGS or sth. But fail to use element id to fetch its geometry...

**Reply by aothms** [reply]
> the output doesnot contains all vertices I need

You take `ifc_file.by_type('IfcBuildingElement')[0]` but I think the issue is simply that the window is not a single product, but many products, so you need to loop over all building elements.

**Reply by sennhvi** [reply]
Thanks for your reply! What confuses me is that the window  is not a single product but it has multiple IFCSHAPEREPRESENTATION to share the same IFCWINDOW, makes it not easy to get geometry data only by its product_id.

**Reply by aothms** [reply]
Yes, but that's a bit of a given in IFC. Depending on the use case you can have either very little or very detailed information on a window. And not always these different aspects of the window are semantically related because modellers predominantly care about the documenting the physical characteristics and not so much delineate on a functional level what a single window constitutes.

**Reply by sennhvi** [reply]
Thanks for your reply, I can understand what you described, so I tried to handle this kind of artifact with multiple shapeRepresentations via geometry iterator, but found that the artifact's shaperepresentation property access during iteration is discontinuous, is this normal?
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/31751605/acc97512-f94d-45ef-8d5d-486e71d8cdd5)
This meant that I couldn't stitch together geometry data that belonged to the same component during iteration

**Reply by sennhvi** [reply]
@aothms Hi, I just found a very useful parameter -- rep, it was previously neglected due to the lack of documentation and insufficient exploration of the source code. 

```
representations = product.Representation.Representations
for rep in representations:
    shape = ifcopenshell.geom.create_shape(settings, product, rep)
```

By setting the third parameter, I can get specific representation's geometry and now this window show perfectly!
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/31751605/1f72f803-a94d-4e60-af74-0c60ea80ac0b)

Thanks again for your reply!

**Reply by aothms** [reply]
> rep, it was previously neglected due to the lack of documentation 

To be honest I also forgot a bit about that. If you can find some time to add a note to the documentation that would be greatly appreciated. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/docs/ifcopenshell-python/geometry_processing.rst#individual-processing

**Reply by sennhvi** [reply]
Yes, I have added a note in  PR #4154 :)
